### ISSUE TITLE
Debug assertion triggered by a simple LR(1) grammar

### ISSUE THREAD

**Issue by osa1** [lalrpop/lalrpop#684]
Reproduces with current master branch (b777a6a).

```
$ cat test.lalrpop
grammar;

extern {
    enum lexer::Tok {
        "a" => Token::A,
        "b" => Token::B,
        "c" => Token::C,
        "d" => Token::D,
        "e" => Token::E,
    }
}

pub Entry: Expr = {
    "a" <x:X> "d" => Expr::E1(x),
    "a" <y:Y> "c" => Expr::E2(y),
    "b" <y:Y> "d" => Expr::E3(y),
    "b" <x:X> "c" => Expr::E4(x),
};

X: usize = {
    "e" <x:X> => x + 1,
    "e" => 1,
};

Y: usize = {
    "e" <y:Y> => y + 1,
    "e" => 1,
};

$ ~/rust/lalrpop/target/debug/lalrpop test.lalrpop
processing file `test.lalrpop`
thread 'main' panicked at 'assertion failed: state.index.0 == index', lalrpop/src/lr1/mod.rs:54:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:48:5
   3: lalrpop::lr1::rewrite_state_indices
             at /home/omer/rust/lalrpop/lalrpop/src/lr1/mod.rs:54:9
   4: lalrpop::lr1::build_states
             at /home/omer/rust/lalrpop/lalrpop/src/lr1/mod.rs:37:5
   5: lalrpop::build::emit_recursive_ascent
             at /home/omer/rust/lalrpop/lalrpop/src/build/mod.rs:386:25
   6: lalrpop::build::process_file_into
             at /home/omer/rust/lalrpop/lalrpop/src/build/mod.rs:134:26
   7: lalrpop::build::process_file
             at /home/omer/rust/lalrpop/lalrpop/src/build/mod.rs:68:5
   8: lalrpop::api::Configuration::process_file
             at /home/omer/rust/lalrpop/lalrpop/src/api/mod.rs:244:9
   9: lalrpop::main1
             at /home/omer/rust/lalrpop/lalrpop/src/main.rs:156:15
  10: lalrpop::main
             at /home/omer/rust/lalrpop/lalrpop/src/main.rs:90:5
  11: core::ops::function::FnOnce::call_once
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by andersk** [comment]
`git bisect` shows that this failure was introduced by cb1924632b44417bfbec47ef533b599698878f07 (#523, cc @Marwes).

**Comment by andersk** [comment]
I thiiink this is the fix (#697):

```diff
--- a/lalrpop/src/lr1/lane_table/construct/merge.rs
+++ b/lalrpop/src/lr1/lane_table/construct/merge.rs
@@ -143,7 +143,8 @@ impl<'m, 'grammar> Merge<'m, 'grammar> {
     fn clone(&mut self, state: StateIndex) -> StateIndex {
         // create a new state with same contents as the old one
         let new_index = StateIndex(self.states.len());
-        let new_state = self.states[state.0].clone();
+        let mut new_state = self.states[state.0].clone();
+        new_state.index = new_index;
         self.states.push(new_state);
 
         // track the original index and clones
```
